datumaro/components/converter.py

```diff
@@ -34,8 +36,31 @@ def convert(cls, extractor, save_dir, **options):
         return converter.apply()
 
     @classmethod
+    @error_rollback
     def patch(cls, dataset, patch, save_dir, **options):
-        return cls.convert(dataset, save_dir, **options)
+        # This solution is not any better in performance than just
+        # writing a dataset, but in case of patching (i.e. writing
+        # to the previous location), it allows to avoid many problems
+        # with removing and replacing existing files. Surely, this
+        # approach also has problems with removal of the given directory.
+        # Problems can occur if we can't remove the directory,
+        # or want to reuse the given directory. It can happen if it
+        # is mounted or (sym-)linked.
+        # Probably, a better solution could be to wipe directory
+        # contents and write new data there. Note that directly doing this
+        # also doesn't work, because images may be needed for writing.
+        tmpdir = mkdtemp(dir=osp.dirname(save_dir),
+            prefix=osp.basename(save_dir), suffix='.tmp')
+        on_error_do(shutil.rmtree, tmpdir, ignore_errors=True)
+
+        retval = cls.convert(dataset, tmpdir, **options)
+
+        if osp.isdir(save_dir):
```

Comment:
I don't think this check should be done here. If it fails, then the directory will be left with `mkdtemp`'s default permissions (700), which would be surprising. Instead, you should do it at the start of the method, and if it fails, then you can just do the conversion normally (without creating a temporary directory).